app/models/attachment_reference.rb

```diff
@@ -0,0 +1,57 @@
+class AttachmentReference < ActiveRecord::Base
+  after_initialize :set_expiry_at, if: :new_record?
+  before_save :update_expiry_at
+
+  belongs_to :attachable, polymorphic: true, inverse_of: :attachment_references
+  belongs_to :attachment, inverse_of: :attachment_references
+
+  delegate :open, :url, :path, to: :attachment
+
+  # Get the name from the file and then further build or find an attachment based on file's SHA256
+  # hash.
+  #
+  # @param [File|ActionDispatch::Http::UploadedFile] The uploaded file.
+  def file=(file)
+    self.name = filename(file)
+    self.attachment = find_or_initialize_attachment_by(file)
+  end
+
+  private
+
+  # Calculates the SHA256 hash and return the existing attachment which has the same hash.
+  # Otherwise a new attachment will be built.
+  #
+  # @param [File|ActionDispatch::Http::UploadedFile] The uploaded file.
+  # @return [Attachment] The attachment which contains the file.
+  def find_or_initialize_attachment_by(file)
+    # Get the actual file by #tempfile if the file is an `ActionDispatch::Http::UploadedFile`.
+    digest = Digest::SHA256.file(file.try(:tempfile) || file).hexdigest
+    Attachment.find_by(name: digest) || build_attachment(name: digest, file_upload: file)
+  end
+
+  # Infer the name of the file.
+  #
+  # @param [File|ActionDispatch::Http::UploadedFile] The uploaded file.
+  # @return [String] The filename.
+  def filename(file)
+    if file.respond_to?(:original_filename)
+      file.original_filename
+    else
+      File.basename(file)
+    end
+  end
+
+  # Clear the expiry time when attachable is present and set the expiry time when attachable is
+  # removed.
+  def update_expiry_at
+    if attachable
+      self.expiry_at = nil
+    else
+      set_expiry_at
+    end
+  end
+
+  def set_expiry_at
+    self.expiry_at ||= 1.day.from_now
```

Comment:
OK, depends on the use case, I explicitly make it not renewable...  will change that.